apps/opentelemetry_experimental/src/otel_aggregation_sum.erl

```
@@ -0,0 +1,95 @@
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc
+%% @end
+%%%-------------------------------------------------------------------------
+-module(otel_aggregation_sum).
+
+-export([init/2,
+         aggregate/3,
+         checkpoint/5,
+         collect/4]).
+
+-include("otel_metrics.hrl").
+-include_lib("opentelemetry_api_experimental/include/otel_metrics.hrl").
+
+-type t() :: #sum_aggregation{}.
+
+-export_type([t/0]).
+
+init(Key, _Options) ->
+    #sum_aggregation{key=Key,
+                     start_time_unix_nano=erlang:system_time(nanosecond),
+                     value=0}.
+
+aggregate(Tab, Key, Value) ->
+    try
+        _ = ets:update_counter(Tab, Key, {#sum_aggregation.value, Value}),
+        true
+    catch
+        error:badarg ->
+            %% the use of `update_counter' guards against conflicting with another process
+            %% doing the update at the same time -- if we ever support that
+
+            %% the default isn't just given in the first `update_counter' because then
+            %% we'd have to call `system_time' for every single measurement taken
+            _ = ets:update_counter(Tab, Key, {#sum_aggregation.value, Value},
+                                   init(Key, [])),
+            true
+    end.
+
+-dialyzer({nowarn_function, checkpoint/5}).
+checkpoint(Tab, Name, ?AGGREGATION_TEMPORALITY_DELTA, ?VALUE_TYPE_INTEGER, CollectionStartNano) ->
+    MS = [{#sum_aggregation{key='$1',
+                            value='$2',
+                            _='_'},
```

I would avoid a wildcard match here; if a field is ever added to the record, this will silently succeed matching but the re-insertion will reset the fields.

By forcing an explicit key match for the patterns, we'll catch changes in tests by having any match failing. OTOH it will be costlier to match though so maybe you're fine with this. If so, maybe pair the record definition with a comment about needing to update the aggregators to care for the fields?

Feel free to disregard this if we expect the spec to be stable.

---

Yea, can't remember why I did this, I think it was just to guard against it not checkpointing if the record was ever changed. Sure it would reset the other fields but at least it would checkpoint the key and value. Not sure that makes sense anymore, so I'll probably change it in this PR.

---

Oh na, now I see, I did that since I did consider the record stable. But I think being explicit will make it less likely to cause confusion at least in the future.